ocrd/ocrd/cli/validate.py

```diff
@@ -0,0 +1,100 @@
+import sys
+
+import click
+from json import loads
+import codecs
+
+from ocrd import Resolver
+from ocrd.task_sequence import ProcessorTask
+
+from ocrd_utils import (
+    parse_json_string_or_file
+)
+from ocrd_validators import (
+    OcrdToolValidator,
+    OcrdZipValidator,
+    PageValidator,
+    ParameterValidator,
+    WorkspaceValidator,
+)
+
+def _inform_of_result(report):
+    if not report.is_valid:
+        print(report.to_xml())
```

Comment:
Maybe this should be done via logging facilities?

Reply:
That report is multiline which isn't ideal for logging. Also, logging can be disabled by the user but if a validation check fails, we always want to show the report.

Reply:
> That report is multiline which isn't ideal for logging.

You could wrap the individual lines of the report into log messages.

> Also, logging can be disabled by the user but if a validation check fails, we always want to show the report.

But isn't this a case of logging-as-output? And you could even argue the actual result is the exit value.

Anyway, I leave that entirely to you – withdrawing my request.

Reply:
Yeah, I think the value of having a clean solution is outweighed by convenience. I'm thinking of users running this in batch on lots of files, generating lots of reports (@tboenig does that regularly for our GT). As-is the reports are crude but well-formed XML which can be used.

Needs more testing though, I'll try to write some unit tests before merging.

Reply:
> As-is the reports are crude but well-formed XML which can be used.

Oh, right, I did not think of this. Then maybe you want to allow writing the reports into files as well (like dinglehopper, perhaps even registered in METS)?

Reply:
> like dinglehopper, perhaps even registered in METS

But the point of this is to have all the invidiual validation methods in one place instead of workspace validation (for which I would consider such a parameter if I there is demand).

> Then maybe you want to allow writing the reports into files as well

Very little advantage over shell redirection I think. Basically I'm trying to avoid @tboenig having to change his setup at all :D

Reply:
> But the point of this is to have all the invidiual validation methods in one place instead of workspace validation (for which I would consider such a parameter if I there is demand).

Sorry, I forgot this is already about probing different parts individually. METS does not apply here.

> Very little advantage over shell redirection I think.

Even with the prospect of separating different XML reports into different files (which redirection cannot do easily)?

> Basically I'm trying to avoid @tboenig having to change his setup at all :D

But this is a new interface, isn't it? The existing workflows with validation should be unrelated...

Reply:
> Even with the prospect of separating different XML reports into different files (which redirection cannot do easily)?

You'd still have to execute once per subcommand so you could redirect everyone of them (not arguing for the sake of arguing here, it's not a problem to implement if I missed a use case).

> But this is a new interface, isn't it? 

Yeah that was just a joke

Reply:
> You'd still have to execute once per subcommand so you could redirect everyone of them

Sorry again, I was too confused to see that simple fact.

> > But this is a new interface, isn't it?
> 
> Yeah that was just a joke

Got it (eventually)!